Fix images being interpreted as code blocks and inline code blocks

MaderHatt3r opened the pull request:
This fixes images failing to be converted when they are in the middle of code blocks or are converted to an inline code block due to formatting.

Mr0grog commented on lib/fix-google-html.js:
```diff
@@ -164,6 +164,13 @@ export function unInlineStyles(node) {
  * @param {HastNodes} node Fix the tree below this node
  */
 function convertInlineStylesToElements(node) {
+  // Helper to check if a node contains any images
+  const containsImage = (node) => {
+    if (node.tagName === 'img') return true;
+    if (!node.children?.length) return false;
+    return node.children.some(child => containsImage(child));
+  };
```
Could you please move this out into a module-level function, that way it doesn’t need to get redefined for every call to `convertinlineStylesToElements`? It also looks like this is identical to the inline function in `createCodeBlocks()`, so pulling this up to the module level means you can just call it from there, too.

Also, I think it might be a good idea to refactor this as `containsReplacedElement` or `containsNonTextContent` or something. I haven't ever tried embedding audio or video in a Google Doc (not even sure if you can?), but I suspect it would be a good future-proof idea to cover any kind of element that represents some other non-text resource to be displayed (broadly referred to as [“replaced elements”](https://developer.mozilla.org/en-US/docs/Glossary/Replaced_elements) in HTML).

MaderHatt3r replied:
Fixed

Mr0grog commented on lib/fix-google-html.js:
```diff
@@ -471,6 +479,11 @@ function isAllTextCode(parent) {
 
   let hasText = false;
   for (const child of parent.children) {
+    // Don't create code blocks if there are images present
+    if (child.tagName === 'img') {
+      return false;
+    }
```
It would probably be good to extract this check and the corresponding one in `containsImage()` into a function and also generalize for replaced or non-text content as mentioned above. For example, `isReplacedElement(node)` or `isNonTextContent(node)` or something along those lines.

MaderHatt3r replied:
fixed

Mr0grog commented on lib/fix-google-html.js:
```diff
@@ -504,27 +517,40 @@ export function createCodeBlocks(node) {
   const codeBlocks = [];
   let activeCodeBlock = null;
+  
+  // Helper to check if a node contains any images
+  const containsImage = (node) => {
+    if (node.tagName === 'img') return true;
+    if (!node.children?.length) return false;
+    return node.children.some(child => containsImage(child));
+  };
```
This is the same as the `containsImage()` function defined inside `convertInlineStylesToElements()`, right? Please just make these one function you can call from both places.

MaderHatt3r replied:
fixed

Mr0grog commented on lib/fix-google-html.js:
```diff
@@ -504,27 +517,40 @@ export function createCodeBlocks(node) {
+      if (isAllTextCode(child) && !containsImage(child)) {
         if (!activeCodeBlock) {
-          activeCodeBlock = { node, start: i, end: 0 };
+          activeCodeBlock = { node, start: i, end: i + 1 };
           codeBlocks.push(activeCodeBlock);
+        } else {
+          activeCodeBlock.end = i + 1;
```
Is there a case the previous logic wasn’t working for? What was this solving?

MaderHatt3r replied:
I think I started playing with this code originally because setting the end seemed inconsistent and never finished my thought process. Good catch. 

I removed my edits and updated the code with inline comments to explain what's happening. I also removed `end: 0` because it seems a little misleading. `end: undefined` might also be acceptable if you want it more explicit. We "could" update this code to `i+1` and remove the lines after the for loop (shown below) which would consistently keep track of the end of the block all within the loop instead of half in and half out, but the assignment per element isn't really necessary even if it's not that expensive of an operation.  

```
if (activeCodeBlock) {
  activeCodeBlock.end = node.children.length;
}
```

I'll change it to whatever you want or change it back to original if you prefer. Just let me know.

Mr0grog replied:
> I also removed `end: 0` because it seems a little misleading. `end: undefined` might also be acceptable

Oh, this is actually a performance optimization, although it only really matters on a page with a lot of monospaced/code text. Under the hood, when a JS engine generates optimized machine code, it figures out how to represent and do operations on your objects based on their shape (i.e. what `(property-name: type-of-value)` pairs they have). Whenever it tries to run that optimized code with an object of a different shape (which will happen here when you are adding a new property or changing the type of the property), it has to back up and re-run the operation with de-optimized code (and then later figure out if it should optimize again with the new shape).

If you think the code would be clearer, a better solution is probably just to initialize `end` to `i + 1` instead of `0`.

MaderHatt3r replied:
That makes a lot of sense. Thanks for taking the time to explain it!

Mr0grog reviewed (changes requested):
Sorry I wasn’t able to take a look at this earlier! I left some notes about pretty minor refactors inline and one question.

Do you have a Google Doc to use as a test case here? I can help hook it up to the tests if it’s not obvious or you don’t have time to do so.

Otherwise, this PR makes sense to me. Will try and get this landed as soon as the above stuff is done. 👍

MaderHatt3r commented:
> Do you have a Google Doc to use as a test case here? I can help hook it up to the tests if it’s not obvious or you don’t have time to do so.

https://docs.google.com/document/d/1rmdBM68XKw5t8QsvoU-3CznFxk515nbFzZ0KVCF1-Cc/edit?usp=sharing

Mr0grog reviewed (commented):
Alright, I added that doc as a fixture and set the tests to run. It passes, but it looks like there are some lint errors in your changes to `lib/fix-google-html.js`.

MaderHatt3r commented:
I finished some projects I was working on and should have time to apply the pr comments and linting soon. Sorry for the delay.

MaderHatt3r commented:
The only lint error I saw was a comment that was too long. I fixed it. I added a lint script to the package.json file. Please let me know if I was running it incorrectly.

Mr0grog commented:
> The only lint error I saw was a comment that was too long. I fixed it. I added a lint script to the package.json file. Please let me know if I was running it incorrectly.

I think you are missing [the Prettier check](https://github.com/Mr0grog/google-docs-to-markdown/blob/e091cc61d6b470ef3f5f43e8c8055f147e29cfdb/.github/workflows/ci.yml#L28-L29). I went ahead and added it here: a3aae93e89f95827f879a278c9dfd85ece7c5a3a

Mr0grog reviewed (approved):
This looks great! I made a few *very minor* performance tweaks, and I think this is probably ready to land.

> Sorry for the delay.

No worries, it’s all done now.
